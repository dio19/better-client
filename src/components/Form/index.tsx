import * as React from "react";
import * as Yup from "yup";
import { Formik } from "formik";
import Grid from "@mui/material/Grid";
import Switch from "@mui/material/Switch";
import Button from "@mui/material/Button";
import TextField from "@mui/material/TextField";
import dayjs, { Dayjs } from "dayjs";
import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { DateTimePicker } from "@mui/x-date-pickers/DateTimePicker";
import ErrorOutlineIcon from "@mui/icons-material/ErrorOutline";

import CustomTextField from "../CustomTextField";
import { CustomerContext } from "../../context";
import { DataCustomer } from "../../types";

import "./index.scss";

const Form = () => {
  const INITIAL_FORM_STATE = {
    first: "",
    last: "",
    email: "",
    company: "",
    country: "",
  };

  const FORM_VALIDATION = Yup.object().shape({
    first: Yup.string().required("Required"),
    last: Yup.string().required("Required"),
    email: Yup.string().email("Invalid email.").required("Required"),
    company: Yup.string().required("Required"),
    country: Yup.string(),
  });

  const { DataContext } = React.useContext(CustomerContext);
  const [ID, setID] = React.useState(0);
  const [checked, setChecked] = React.useState(false);
  const [customerById, setCustomerById] = React.useState<
    DataCustomer | undefined
  >(undefined);
  const [idNotFound, setIdNotFound] = React.useState(false);
  const [method, setMethod] = React.useState(false);
  const [valueDate, setValueDate] = React.useState<Dayjs | null>(dayjs());

  const handleChangeMethod = (event: React.ChangeEvent<HTMLInputElement>) => {
    setMethod(event.target.checked);
  };

  const handleChangeDate = (newValue: Dayjs | null) => {
    setValueDate(newValue);
  };

  const handleChangeCheck = (e: React.FormEvent<HTMLInputElement>) => {
    const target = e.target as HTMLInputElement;
    setID(Number(target.value));
  };

  const handleKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    if (e.key === "Enter" && ID !== 0) {
      const customer = DataContext.find((customer) => {
        return customer.id === ID;
      });

      if (customer !== undefined) {
        setCustomerById({ ...customer });
        setChecked(true);
      } else {
        setIdNotFound(true);
        setChecked(false);
      }
    }
  };

  return (
    <div className="form">
      <div className="form__check">
        <div className="input">
          <input
            className="input-container"
            type="number"
            onChange={handleChangeCheck}
            onKeyDown={handleKeyDown}
          />
          <span className="input-label">CHECK ID</span>
        </div>
        <div className="form__check__not-found">
          {idNotFound && !checked && (
            <>
              <ErrorOutlineIcon color="error" />
              <div className="form__check__not-found-text">ERROR</div>
            </>
          )}
        </div>
        {checked && (
          <div className="form__check__method">
            <Switch checked={method} onChange={handleChangeMethod} />
            <div className="form__check__method-label">
              {method ? "DELETE" : "PUT"}
            </div>
          </div>
        )}
      </div>
      <Formik
        initialValues={{
          ...INITIAL_FORM_STATE,
        }}
        validationSchema={FORM_VALIDATION}
        onSubmit={(values) => {
          console.log({ ...values, created_at: dayjs(valueDate).toString() });
        }}
      >
        {({ values, errors, touched, handleChange, handleSubmit }) => (
          <form onSubmit={handleSubmit}>
            <Grid
              container
              rowSpacing={2}
              columnSpacing={{ xs: 1, sm: 2, md: 3 }}
            >
              <Grid item md={6}>
                <CustomTextField
                  id="first"
                  value={values.first}
                  label="Firstname"
                  onChange={handleChange}
                  error={touched.first && Boolean(errors.first)}
                  helperText={
                    customerById !== undefined
                      ? customerById.first
                      : touched.first && errors.first
                  }
                  disabled={!checked || method}
                />
              </Grid>
              <Grid item md={6}>
                <CustomTextField
                  id="last"
                  value={values.last}
                  label="Lastname"
                  onChange={handleChange}
                  error={touched.last && Boolean(errors.last)}
                  helperText={
                    customerById !== undefined
                      ? customerById.last
                      : touched.last && errors.last
                  }
                  disabled={!checked || method}
                />
              </Grid>
              <Grid item md={6}>
                <CustomTextField
                  id="email"
                  value={values.email}
                  label="Email"
                  onChange={handleChange}
                  error={touched.email && Boolean(errors.email)}
                  helperText={
                    customerById !== undefined && !errors.email
                      ? customerById.email
                      : touched.email && errors.email
                  }
                  disabled={!checked || method}
                />
              </Grid>
              <Grid item md={6}>
                <CustomTextField
                  id="company"
                  value={values.company}
                  label="Company"
                  onChange={handleChange}
                  error={touched.company && Boolean(errors.company)}
                  helperText={
                    customerById !== undefined
                      ? customerById.company
                      : touched.company && errors.company
                  }
                  disabled={!checked || method}
                />
              </Grid>
              <Grid item md={6}>
                <CustomTextField
                  id="country"
                  value={values.country}
                  label="Country"
                  onChange={handleChange}
                  error={touched.country && Boolean(errors.country)}
                  helperText={
                    customerById !== undefined
                      ? customerById.country
                      : touched.country && errors.country
                  }
                  disabled={!checked || method}
                />
              </Grid>
              <Grid item md={6}>
                <LocalizationProvider dateAdapter={AdapterDayjs}>
                  <DateTimePicker
                    label="Created at"
                    value={valueDate}
                    onChange={handleChangeDate}
                    disabled={!checked || method}
                    renderInput={(params) => (
                      <TextField
                        helperText={
                          customerById !== undefined && customerById.created_at
                        }
                        {...params}
                      />
                    )}
                  />
                </LocalizationProvider>
              </Grid>
            </Grid>
            <div className="form-post__button">
              <Button
                type="submit"
                variant="contained"
                color="warning"
                size="small"
                disabled={!checked}
              >
                {method ? "Delete Customer" : "Update Customer"}
              </Button>
            </div>
          </form>
        )}
      </Formik>
    </div>
  );
};

export default Form;
