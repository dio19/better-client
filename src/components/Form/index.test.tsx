import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import Form from "./index";

describe("Form Component", () => {
  test("It should render Form Component correctly", () => {
    const { asFragment } = render(
      <Router>
        <Form />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
