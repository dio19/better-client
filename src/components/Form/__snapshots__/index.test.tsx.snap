// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Component It should render Form Component correctly 1`] = `
<DocumentFragment>
  <div
    class="form"
  >
    <div
      class="form__check"
    >
      <div
        class="input"
      >
        <input
          class="input-container"
          type="number"
        />
        <span
          class="input-label"
        >
          CHECK ID
        </span>
      </div>
      <div
        class="form__check__not-found"
      />
    </div>
    <form>
      <div
        class="MuiGrid-root MuiGrid-container css-1v2c24a-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorWarning Mui-disabled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-mk9tez-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="first"
              id="first-label"
            >
              Firstname
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorWarning Mui-disabled MuiInputBase-formControl css-11qfo8z-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled=""
                id="first"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Firstname *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorWarning Mui-disabled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-mk9tez-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="last"
              id="last-label"
            >
              Lastname
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorWarning Mui-disabled MuiInputBase-formControl css-11qfo8z-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled=""
                id="last"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Lastname *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorWarning Mui-disabled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-mk9tez-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="email"
              id="email-label"
            >
              Email
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorWarning Mui-disabled MuiInputBase-formControl css-11qfo8z-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled=""
                id="email"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Email *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorWarning Mui-disabled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-mk9tez-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="company"
              id="company-label"
            >
              Company
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorWarning Mui-disabled MuiInputBase-formControl css-11qfo8z-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled=""
                id="company"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Company *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorWarning Mui-disabled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-mk9tez-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="country"
              id="country-label"
            >
              Country
              <span
                aria-hidden="true"
                class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
              >
                 *
              </span>
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorWarning Mui-disabled MuiInputBase-formControl css-11qfo8z-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                disabled=""
                id="country"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Country *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 css-1k7sk4d-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r5:"
              id=":r5:-label"
            >
              Created at
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                disabled=""
                id=":r5:"
                placeholder="mm/dd/yyyy hh:mm (a|p)m"
                type="tel"
                value="10/31/2022 11:18 AM"
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
              >
                <button
                  aria-label="Choose date, selected date is Oct 31, 2022"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CalendarIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                    />
                  </svg>
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    Created at
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="form-post__button"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedWarning MuiButton-sizeSmall MuiButton-containedSizeSmall Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedWarning MuiButton-sizeSmall MuiButton-containedSizeSmall css-1edgfgr-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="submit"
        >
          Update Customer
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
